// react-pouchdb/node/cjs/BaseComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _getPrototypeOf = _interopRequireDefault(require("@babel/runtime/core-js/object/get-prototype-of"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = require("react");

var _shouldUpdate = _interopRequireDefault(require("./shouldUpdate"));

var _renderProps = require("./renderProps");

var _withDB = require("./withDB");

var BaseComponent =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(BaseComponent, _Component);

  function BaseComponent() {
    (0, _classCallCheck2.default)(this, BaseComponent);
    return (0, _possibleConstructorReturn2.default)(this, (BaseComponent.__proto__ || (0, _getPrototypeOf.default)(BaseComponent)).apply(this, arguments));
  }

  (0, _createClass2.default)(BaseComponent, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this._isMounted = true;

      this._listen();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if ((0, _shouldUpdate.default)(prevProps, this.props)) {
        this.unlisten();

        this._listen();
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      this._isMounted = false;
      this.unlisten();
    }
  }, {
    key: "setStateIfMounted",
    value: function setStateIfMounted(state) {
      if (!this._isMounted) {
        return;
      }

      this.setState(state);
      return true;
    }
  }, {
    key: "_listen",
    value: function _listen() {
      this.listen((0, _renderProps.omit)(this.props));
    }
  }, {
    key: "unlisten",
    value: function unlisten() {
      var cancel = this.cancel;

      if (cancel) {
        cancel();
      }
    }
  }]);
  return BaseComponent;
}(_react.Component);

exports.default = BaseComponent;
Object.defineProperty(BaseComponent, "propTypes", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: (0, _objectSpread2.default)({}, _renderProps.propTypes, _withDB.propTypes)
});
