// react-pouchdb/node/cjs/withDB.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.propTypes = void 0;

var _react = _interopRequireWildcard(require("react"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _assign = _interopRequireDefault(require("@babel/runtime/core-js/object/assign"));

var _pouchdbNode = _interopRequireDefault(require("pouchdb-node"));

var _propTypes = require("prop-types");

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _DBContext = require("./DBContext");

var propTypes = {
  db: (0, _propTypes.instanceOf)(_pouchdbNode.default).isRequired
};
exports.propTypes = propTypes;

var withDB = function withDB(Component) {
  return (0, _react.forwardRef)((0, _hoistNonReactStatics.default)((0, _assign.default)(function (props, ref) {
    return _react.default.createElement(_DBContext.Consumer, null, function (db) {
      return _react.default.createElement(Component, (0, _extends2.default)({}, props, {
        db: db,
        ref: ref
      }));
    });
  }, {
    displayName: "withDB(".concat(Component.displayName || Component.name, ")"),
    WrappedComponent: Component
  }), Component));
};

var _default = withDB;
exports.default = _default;
