// react-pouchdb/browser/cjs/PouchDB.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _getPrototypeOf = _interopRequireDefault(require("@babel/runtime/core-js/object/get-prototype-of"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _propTypes = require("prop-types");

var _pouchdbConnections = require("./pouchdbConnections");

var _DBContext = require("./DBContext");

var PouchDB =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(PouchDB, _Component);

  function PouchDB(props) {
    var _this;

    (0, _classCallCheck2.default)(this, PouchDB);
    _this = (0, _possibleConstructorReturn2.default)(this, (PouchDB.__proto__ || (0, _getPrototypeOf.default)(PouchDB)).call(this, props));
    var children = props.children,
        maxListeners = props.maxListeners,
        options = (0, _objectWithoutProperties2.default)(props, ["children", "maxListeners"]);
    _this.db = (0, _pouchdbConnections.create)(options);

    if (maxListeners) {
      _this.db.setMaxListeners(maxListeners);
    }

    return _this;
  }

  (0, _createClass2.default)(PouchDB, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var _props = this.props,
          children = _props.children,
          maxListeners = _props.maxListeners,
          options = (0, _objectWithoutProperties2.default)(_props, ["children", "maxListeners"]);
      (0, _pouchdbConnections.close)(options);
    }
  }, {
    key: "render",
    value: function render() {
      return _react.default.createElement(_DBContext.Provider, {
        value: this.db
      }, this.props.children);
    }
  }]);
  return PouchDB;
}(_react.Component);

exports.default = PouchDB;
Object.defineProperty(PouchDB, "propTypes", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: {
    children: _propTypes.node,
    maxListeners: _propTypes.number,
    name: _propTypes.string.isRequired
  }
});
Object.defineProperty(PouchDB, "defaultProps", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: {
    children: null,
    maxListeners: undefined
  }
});
