// react-pouchdb/es/PouchDB.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _Object$getPrototypeOf from "@babel/runtime/core-js/object/get-prototype-of";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _inherits from "@babel/runtime/helpers/inherits";
import React from "react";
import { Component } from 'react';
import { node, number, string } from 'prop-types';
import { create, close } from './pouchdbConnections';
import { Provider } from './DBContext';

var PouchDB =
/*#__PURE__*/
function (_Component) {
  _inherits(PouchDB, _Component);

  function PouchDB(props) {
    var _this;

    _classCallCheck(this, PouchDB);

    _this = _possibleConstructorReturn(this, (PouchDB.__proto__ || _Object$getPrototypeOf(PouchDB)).call(this, props));

    var children = props.children,
        maxListeners = props.maxListeners,
        options = _objectWithoutProperties(props, ["children", "maxListeners"]);

    _this.db = create(options);

    if (maxListeners) {
      _this.db.setMaxListeners(maxListeners);
    }

    return _this;
  }

  _createClass(PouchDB, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var _props = this.props,
          children = _props.children,
          maxListeners = _props.maxListeners,
          options = _objectWithoutProperties(_props, ["children", "maxListeners"]);

      close(options);
    }
  }, {
    key: "render",
    value: function render() {
      return React.createElement(Provider, {
        value: this.db
      }, this.props.children);
    }
  }]);

  return PouchDB;
}(Component);

Object.defineProperty(PouchDB, "propTypes", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: {
    children: node,
    maxListeners: number,
    name: string.isRequired
  }
});
Object.defineProperty(PouchDB, "defaultProps", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: {
    children: null,
    maxListeners: undefined
  }
});
export { PouchDB as default };
