// react-pouchdb/cjs/Get.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _keys = _interopRequireDefault(require("@babel/runtime/core-js/object/keys"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _getPrototypeOf = _interopRequireDefault(require("@babel/runtime/core-js/object/get-prototype-of"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _propTypes = require("prop-types");

var _BaseComponent2 = _interopRequireDefault(require("./BaseComponent"));

var _attachmentsAsUint8Arrays = _interopRequireDefault(require("./attachmentsAsUint8Arrays"));

var _changesCache = _interopRequireDefault(require("./changesCache"));

var _renderProps = _interopRequireDefault(require("./renderProps"));

var _withDB = _interopRequireDefault(require("./withDB"));

var UINT8ARRAY = 'u8a';
var ALLOWED_LIVE_OPTIONS = ['attachments', 'ajax', 'binary', 'id'];

var Get =
/*#__PURE__*/
function (_BaseComponent) {
  (0, _inherits2.default)(Get, _BaseComponent);

  function Get() {
    var _ref;

    var _temp, _this;

    (0, _classCallCheck2.default)(this, Get);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return (0, _possibleConstructorReturn2.default)(_this, (_temp = _this = (0, _possibleConstructorReturn2.default)(this, (_ref = Get.__proto__ || (0, _getPrototypeOf.default)(Get)).call.apply(_ref, [this].concat(args))), Object.defineProperty((0, _assertThisInitialized2.default)(_this), "state", {
      configurable: true,
      enumerable: true,
      writable: true,
      value: {}
    }), _temp));
  }

  (0, _createClass2.default)(Get, [{
    key: "listen",
    value: function () {
      var _listen = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2(options) {
        var _this2 = this;

        var db, id, attachments, otherOptions, optionsWithAttachmentAndBinaryOption;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                db = this.props.db;
                id = options.id, attachments = options.attachments, otherOptions = (0, _objectWithoutProperties2.default)(options, ["id", "attachments"]);
                optionsWithAttachmentAndBinaryOption = (0, _objectSpread2.default)({
                  binary: attachments === UINT8ARRAY
                }, otherOptions, {
                  attachments: !!attachments
                });
                _context2.t0 = this;
                _context2.next = 6;
                return this.get(id, optionsWithAttachmentAndBinaryOption);

              case 6:
                _context2.t1 = _context2.sent;

                if (!_context2.t0.setStateIfMounted.call(_context2.t0, _context2.t1)) {
                  _context2.next = 9;
                  break;
                }

                // Live?
                if ((0, _keys.default)(options).every(function (option) {
                  return ALLOWED_LIVE_OPTIONS.includes(option);
                })) {
                  this.cancel = _changesCache.default.call(db, (0, _objectSpread2.default)({}, optionsWithAttachmentAndBinaryOption, {
                    live: true,
                    include_docs: true,
                    since: 'now',
                    doc_ids: [id]
                  }),
                  /*#__PURE__*/
                  function () {
                    var _ref3 = (0, _asyncToGenerator2.default)(
                    /*#__PURE__*/
                    _regenerator.default.mark(function _callee(_ref2) {
                      var doc;
                      return _regenerator.default.wrap(function _callee$(_context) {
                        while (1) {
                          switch (_context.prev = _context.next) {
                            case 0:
                              doc = _ref2.doc;
                              _context.t0 = _this2;
                              _context.next = 4;
                              return _this2.nextState(doc);

                            case 4:
                              _context.t1 = _context.sent;
                              return _context.abrupt("return", _context.t0.setStateIfMounted.call(_context.t0, _context.t1));

                            case 6:
                            case "end":
                              return _context.stop();
                          }
                        }
                      }, _callee, this);
                    }));

                    return function (_x2) {
                      return _ref3.apply(this, arguments);
                    };
                  }());
                }

              case 9:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function listen(_x) {
        return _listen.apply(this, arguments);
      };
    }()
  }, {
    key: "get",
    value: function () {
      var _get = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee3(id, options) {
        var db;
        return _regenerator.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                db = this.props.db;
                _context3.prev = 1;
                _context3.t0 = this;
                _context3.next = 5;
                return db.get(id, options);

              case 5:
                _context3.t1 = _context3.sent;
                _context3.next = 8;
                return _context3.t0.nextState.call(_context3.t0, _context3.t1);

              case 8:
                return _context3.abrupt("return", _context3.sent);

              case 11:
                _context3.prev = 11;
                _context3.t2 = _context3["catch"](1);
                return _context3.abrupt("return", {
                  exists: false
                });

              case 14:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this, [[1, 11]]);
      }));

      return function get(_x3, _x4) {
        return _get.apply(this, arguments);
      };
    }()
  }, {
    key: "nextState",
    value: function () {
      var _nextState = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee4(doc) {
        return _regenerator.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                if (!(this.props.attachments === UINT8ARRAY)) {
                  _context4.next = 6;
                  break;
                }

                _context4.next = 3;
                return (0, _attachmentsAsUint8Arrays.default)(doc._attachments);

              case 3:
                _context4.t0 = _context4.sent;
                _context4.next = 7;
                break;

              case 6:
                _context4.t0 = doc._attachments;

              case 7:
                _context4.t1 = _context4.t0;
                _context4.t2 = doc;
                _context4.t3 = !doc._deleted;
                return _context4.abrupt("return", {
                  attachments: _context4.t1,
                  doc: _context4.t2,
                  exists: _context4.t3
                });

              case 11:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      return function nextState(_x5) {
        return _nextState.apply(this, arguments);
      };
    }()
  }, {
    key: "render",
    value: function render() {
      var _props = this.props,
          db = _props.db,
          otherProps = (0, _objectWithoutProperties2.default)(_props, ["db"]),
          state = this.state,
          exists = this.state.exists;
      return (0, _renderProps.default)(otherProps, exists, (0, _objectSpread2.default)({
        db: db
      }, state));
    }
  }]);
  return Get;
}(_BaseComponent2.default);

Object.defineProperty(Get, "propTypes", {
  configurable: true,
  enumerable: true,
  writable: true,
  value: (0, _objectSpread2.default)({}, _BaseComponent2.default.propTypes, {
    attachments: (0, _propTypes.oneOf)([true, false, UINT8ARRAY]),
    binary: _propTypes.bool,
    id: _propTypes.string.isRequired
  })
});

var _default = (0, _withDB.default)(Get);

exports.default = _default;
